src/Microsoft.Data.SqlClient/add-ons/AzureKeyVaultProvider/SqlColumnEncryptionAzureKeyVaultProvider.cs

```diff
@@ -282,173 +276,69 @@ public override byte[] DecryptColumnEncryptionKey(string masterKeyPath, string e
+            ValidateNotEmpty(columnEncryptionKey, nameof(columnEncryptionKey));
 
-            // Validate encryptionAlgorithm
-            this.ValidateEncryptionAlgorithm(ref encryptionAlgorithm, isSystemOp: false);
-
-            // Validate whether the key is RSA one or not and then get the key size
-            int keySizeInBytes = GetAKVKeySize(masterKeyPath);
+            // Also validates whether the key is RSA one or not and then get the key size
+            KeyCryptographer.AddKey(masterKeyPath);
+            int keySizeInBytes = KeyCryptographer.GetKeySize(masterKeyPath);
 
             // Construct the encryptedColumnEncryptionKey
             // Format is 
             //          version + keyPathLength + ciphertextLength + ciphertext + keyPath + signature
             //
             // We currently only support one version
-            byte[] version = new byte[] { firstVersion[0] };
+            byte[] version = new byte[] { s_firstVersion[0] };
 
             // Get the Unicode encoded bytes of cultureinvariant lower case masterKeyPath
             byte[] masterKeyPathBytes = Encoding.Unicode.GetBytes(masterKeyPath.ToLowerInvariant());
-            byte[] keyPathLength = BitConverter.GetBytes((Int16)masterKeyPathBytes.Length);
+            byte[] keyPathLength = BitConverter.GetBytes((short)masterKeyPathBytes.Length);
 
             // Encrypt the plain text
-            byte[] cipherText = this.AzureKeyVaultWrap(masterKeyPath, encryptionAlgorithm, columnEncryptionKey);
-            byte[] cipherTextLength = BitConverter.GetBytes((Int16)cipherText.Length);
+            byte[] cipherText = KeyCryptographer.WrapKey(s_keyWrapAlgorithm, columnEncryptionKey, masterKeyPath);
+            byte[] cipherTextLength = BitConverter.GetBytes((short)cipherText.Length);
 
             if (cipherText.Length != keySizeInBytes)
             {
-                throw new CryptographicException(Strings.CiphertextLengthMismatch);
+                throw new CryptographicException(Strings.CipherTextLengthMismatch);
             }
 
             // Compute hash
             // SHA-2-256(version + keyPathLength + ciphertextLength + keyPath + ciphertext) 
-            byte[] hash;
-            using (SHA256 sha256 = SHA256.Create())
-            {
-                sha256.TransformBlock(version, 0, version.Length, version, 0);
-                sha256.TransformBlock(keyPathLength, 0, keyPathLength.Length, keyPathLength, 0);
-                sha256.TransformBlock(cipherTextLength, 0, cipherTextLength.Length, cipherTextLength, 0);
-                sha256.TransformBlock(masterKeyPathBytes, 0, masterKeyPathBytes.Length, masterKeyPathBytes, 0);
-                sha256.TransformFinalBlock(cipherText, 0, cipherText.Length);
-                hash = sha256.Hash;
-            }
+            byte[] hash = version.Concat(keyPathLength).Concat(cipherTextLength).Concat(masterKeyPathBytes).Concat(cipherText).ToArray();
```

Comment:
This is still the preimage at this point and not a hash yet since keyvault does the hashing now. The new library makes this variable name misleading now.